lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -2784,4 +2784,212 @@ bool operator==(const Teids& lhs, const Teids& rhs) {
+      RuleToProcess to_process =
+          insert_dynamic_5g_rule(dynamic_rule, lifetime, session_uc);
+      pending_activation->push_back(to_process);
+    }
+  }
+}
+
+RuleToProcess SessionState::activate_static_5g_rule(
+    const std::string& rule_id, RuleLifetime& lifetime,
+    SessionStateUpdateCriteria* update_criteria) {
+  RuleToProcess to_process;
+  PolicyRule rule;
+  static_rules_.get_rule(rule_id, &rule);
+  rule_lifetimes_[rule_id] = lifetime;
+  active_static_rules_.push_back(rule_id);
+  if (update_criteria) {
+    update_criteria->static_rules_to_install.insert(rule_id);
+    update_criteria->new_rule_lifetimes[rule_id] = lifetime;
+  }
+  increment_rule_stats(rule_id, update_criteria);
+  to_process.version = get_current_rule_version(rule_id);
+  to_process.rule    = rule;
+  to_process.teids   = config_.common_context.teids();
+  return to_process;
+}
+
+optional<RuleToProcess> SessionState::deactivate_static_5g_rule(
+    const std::string& rule_id, SessionStateUpdateCriteria* update_criteria) {
+  RuleToProcess to_process;
+  auto it = std::find(
+      active_static_rules_.begin(), active_static_rules_.end(), rule_id);
+  if (it == active_static_rules_.end()) {
+    return {};
+  }
+  update_criteria->static_rules_to_uninstall.insert(rule_id);
+  active_static_rules_.erase(it);
+  increment_rule_stats(rule_id, update_criteria);
+
+  PolicyRule rule;
+  if (!static_rules_.get_rule(rule_id, &rule)) {
+    rule.set_id(rule_id);
+  }
+
+  to_process.version = get_current_rule_version(rule.id());
+  to_process.rule    = rule;
+  to_process.teids   = config_.common_context.teids();
+  return to_process;
+}
+
+RuleToProcess SessionState::insert_dynamic_5g_rule(
+    const PolicyRule& rule, const RuleLifetime& lifetime,
+    SessionStateUpdateCriteria* session_uc) {
+  RuleToProcess to_process;
+  rule_lifetimes_[rule.id()] = lifetime;
+  dynamic_rules_.insert_rule(rule);
+  if (session_uc) {
+    session_uc->dynamic_rules_to_install.push_back(rule);
+    session_uc->new_rule_lifetimes[rule.id()] = lifetime;
+  }
+  increment_rule_stats(rule.id(), session_uc);
+
+  to_process.version = get_current_rule_version(rule.id());
+  to_process.rule    = rule;
+  to_process.teids   = config_.common_context.teids();
+
+  return to_process;
+}
+
+optional<RuleToProcess> SessionState::remove_dynamic_5g_rule(
+    const std::string& rule_id, PolicyRule* rule_out,
+    SessionStateUpdateCriteria* session_uc) {
+  PolicyRule rule;
+  RuleToProcess to_process;
+  bool removed = dynamic_rules_.remove_rule(rule_id, &rule);
+  if (!removed) {
+    return {};
+  }
+  if (rule_out) {
+    *rule_out = rule;
+  }
+  if (session_uc) {
+    session_uc->dynamic_rules_to_uninstall.insert(rule_id);
+  }
+  increment_rule_stats(rule_id, session_uc);
+
+  to_process.version = get_current_rule_version(rule.id());
+  to_process.rule    = rule;
+  to_process.teids   = config_.common_context.teids();
+
+  return to_process;
+}
```

Comment:
This function is exactly as `remove_dynamic_rule`

Why are we duplicating code?

Reply:
dedicated_bearer and bearer_id_by_policy are not supported in 5G.

Reply:
what about this 

http://github.com/magma/magma/blob/812f351315ac14e614593103956ab35926067bf4/lte/gateway/c/session_manager/SessionState.cpp#L1292-L1292

Reply:
Yes, teid value is not used for 5G. I will remove it.

Reply:
why is `bearer_id_by_policy` are not supported in 5g? Is 5g missing the concept of bearer or policy id?
when is `bearer_id_by_policy` used?
What is used for 5g? 

Can we make the function take both so it can be used either or instead of having to copy paste all those lines?

Reply:
Note i didn't mean that TEID value was not necessary, what I was saying is 4g `activate_static_rule` + `make_rule_to_process` seems to be doing the same as `activate_static_5g_rule`.

If in 5g dedicated bearer does not exist, the code related dedicated bearer and policy id at `make_rule_to_process` will be transparent.

Reply:
Now, We are reused the existing code and added the condition for whether call is 5G or not.